core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java

```diff
@@ -337,10 +339,20 @@ void mergeWith(
 
     // merge subsets
     for (RelSubset otherSubset : otherSet.subsets) {
-      RelSubset subset =
-          getOrCreateSubset(
-              otherSubset.getCluster(),
-              otherSubset.getTraitSet());
+      RelSubset subset = null;
+      RelTraitSet otherTraits = otherSubset.getTraitSet();
+
+      // if it is logical or derived physical traitSet
+      if (otherSubset.state == 0 || otherSubset.isDerived()) {
+        subset = getOrCreateSubset(cluster, otherTraits, false);
+      }
+
+      // it may be required only, or both derived and required,
+      // in which case, register again.
+      if (otherSubset.isRequired()) {
```

Comment:
The first character should be upper case?

Reply:
fixed, thanks.